### ISSUE TITLE
Macro `item` fragment lumps item component tokens together so they can't be parsed separately

### ISSUE THREAD

**Issue by abonander** [rust-lang/rust#38442]
Simplest reproduction ([playpen](https://is.gd/xcMccO)):

```rust
macro_rules! takes_item {
    ($item:item) => (takes_fn!{ $item });
}

macro_rules! takes_fn {
    (fn $fnname:ident() {}) => ();
}

takes_item! {
    fn foo() {}
}
```

Yields the following, very confusing error:
```
error: no rules expected the token `fn foo() { }`
 --> <anon>:2:33
  |
2 |     ($item:item) => (takes_fn!{ $item });
  |          
```

This makes it very difficult to use the `item` fragment type to accept valid function items and then parse them out in nested macros.

For a more detailed use-case, see [this file](https://github.com/abonander/anterofit/blob/item_bug/src/macros/mod.rs).

I'm trying to use nested macros to make parsing optional generics and `where` clauses a lot less clunky, but since the `item` fragment lumps the tokens together, the `method_proto!` macro can't parse the individual tokens anymore.

**Comment by jseyfried** [comment]
This will be fixed in macros by example 2.0 (`macro`).

I think it would be a good idea to also fix this for `macro_rules!`, although we will probably have to make compromises for backwards compatibility (c.f. the example from #26361).

**Comment by jseyfried** [comment]
Closing in favor of #26361.
